### PR TITLE
Change protocol from cloudpickle to msgpack

### PR DESCRIPTION
**Pull request by mrocklin**
I've been changing distributed from cloudpickle to msgpack for message serialization (though keeping cloudpickle for function and data serialization for tasks).  This has been for a few reasons

1.  Cloudpickle *might* be a bottleneck (profiling is hard at this stage)
2.  Cloudpickle introduces some security issues in the scheduler
3.  Msgpack allows us, in the future, to use the scheduler for other languages, like R

Unfortunately msgpack understands less of Python than cloudpickle does, so this change was expensive to implement and messes things up a bit.  For example, msgpack coreces strings to bytestrings, so everywhere in the code that said this `d['status']` now says  this `d[b'status']` and other bit like that.

The transformation is mostly done.  It does not seem to speed up benchmarks considerably, so our bet that this was a bottleneck didn't pay off.  It does make us seem more serious and opens up other language client/workers, which is good.  It will slow down development for the near future though as we pick up lingering bugs from the change and teach ourselves to only use constructs that can flow through msgpack well.

My plan is to see how much I can clean this up for the next half day and then make a decision.  I would welcome thoughts from people that have experience with virtues of various message protocols.  Should we be doing something else?

### CODE REVIEW

**Comment by mrocklin** [comment]
OK, I've take this as an opportunity to start an actual protocol, with separate header and payload.  This also gives us the freedom to clean up some things with msgpack string handling.  This now feels marginally cleaner than what we had before.

File: distributed/protocol.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+**Payload**
+
+The payload is any msgpack serializable value.  It may be compressed based
+on the header.
+
+**Sentinel**
+
+We often terminate each message with a sentinel value.  This happens
+outside of this module though and is not baked in.
```
Comment (mrocklin) [inline distributed/protocol.py]: @quasiben @minrk @broxtronix if any of you have a moment I'd appreciate a quick sanity check on this start of a protocol for `distributed`.

I expect the header to grow more complex in the future (including versions, options to forward, and whatnot.)  For now though this is as simple as I can make it.  Am I missing anything critical here?

**Comment by minrk** [comment]
>  For example, msgpack coerces strings to bytestrings, so everywhere in the code that said this d['status'] now says this d[b'status'] and other bit like that.

This is solvable, it's just a frustrating default behavior (msgpack/msgpack-python#99). You can add `use_bin_type=True` on serialization and `encoding="utf8"` on deserialization, and you will get the str=unicode behavior that's logical for Python.

For an IPython parallel reference, we use two forms of serialization:

1. JSON/msgpack/anything that can handle JSONable data for the message protocol itself
2. pickle (or cloudpickle/dill) for *user*-specified objects (arguments, functions, return types)

One thing I do like about the serialization in IPython parallel is that it's pluggable. I can switch from JSON to pickle to msgpack with a line or two of config. At a high level, the 'protocol' is that a message is a JSON-able dictionary. And the protocol implementation (`jupyter_client.Session`) has a packer and unpacker, which are callables that conform to the interface:

- `packer(msg_dict) => bytes`
- `unpacker(packer(msg_dict)) == msg_dict`

Connection files, which are what include port/auth info for connecting to the Schedulers, also include references to the packer/unpacker in use. That makes it easy to compare the performance of msgpack vs json vs pickle for serialization.

**Comment by mrocklin** [comment]
> This is solvable, it's just a frustrating default behavior (msgpack/msgpack-python#99). You can add use_bin_type=True on serialization and encoding="utf8" on deserialization, and you will get the str=unicode behavior that's logical for Python.

Wow.  I really should have spent more time googling this earlier.  That's super helpful.  Thanks!

**Comment by mrocklin** [comment]
OK, things are looking good on this.  Planning to merge soon.  Thanks all for the feedback.

File: distributed/center.py [hunk 2]
```diff
@@ -108,21 +109,23 @@ def port(self):
...
         self.worker_services[address] = services
         logger.info("Register %s", str(address))
-        return b'OK'
+        return 'OK'
 
     def unregister(self, stream, address=None):
+        address = coerce_to_address(address)
         if address not in self.has_what:
-            return b'Address not found: ' + str(address).encode()
+            return 'Address not found: ' + str(address).encode()
```
Comment (minrk) [inline distributed/center.py]: Is this `.encode()` intended, or should it just be `str(address)`?

Comment (mrocklin) [inline distributed/center.py]: It should actually just be `address`.  In the zeal of refactoring I ended up crippling my testing framework.  Working through this and similar issues that snuck by now.

Comment (mrocklin) [inline distributed/center.py]: Fixed

**Comment by mrocklin** [comment]
Have you found any benefit to supporting both JSON and msgpack (and potentially others?)  Has human readability been a significant boon?

**Comment by mrocklin** [comment]
OK, merging for real now.

**Comment by minrk** [comment]
We use JSON (by default, even) for two reasons:

1. it's in the Python stdlib, and early on that mattered to us (we mind dependencies much less, now).
2. ipyparallel sits on the Jupyter protocol, which wants to make it easy for other languages to implement the protocol, and ~everybody has access to a JSON implementation.

So the main reason for the ability to swap serializers is to allow things like IPython parallel to use more performant serializers like msgpack.

**Comment by ogrisel** [comment]
@mrocklin have you made some benchmarks? What is the impact of using msgpack vs pickle?

**Comment by mrocklin** [comment]
Once I switched from the `profile` module to `plop` I found that neither was a significant bottleneck for message serialization.  

Here is a trivial benchmark though:

```python
In [1]: import cloudpickle

In [2]: from distributed.protocol import dumps, loads

In [3]: d = {'a': 1, 'b': 2, 'c': b'234f09shfhfsh', 'd': 'hgfhskfjh'}

In [4]: %timeit cloudpickle.loads(cloudpickle.dumps(d))
The slowest run took 15.43 times longer than the fastest. This could mean that an intermediate result is being cached 
10000 loops, best of 3: 66.5 µs per loop

In [5]: %timeit loads(dumps(d))
The slowest run took 11.09 times longer than the fastest. This could mean that an intermediate result is being cached 
100000 loops, best of 3: 5.49 µs per loop

In [6]: %time _ = [cloudpickle.loads(cloudpickle.dumps(d)) for i in range(1000)]
CPU times: user 70.4 ms, sys: 4.01 ms, total: 74.4 ms
Wall time: 73.8 ms

In [7]: %time _ = [loads(dumps(d)) for i in range(1000)]
CPU times: user 7.68 ms, sys: 0 ns, total: 7.68 ms
Wall time: 7.66 ms
```

**Comment by ogrisel** [comment]
Alright.

**Pull request merged**